### ISSUE TITLE
push() onto const Vec is doing nothing, no warning either

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#64854]
````
rustc 1.39.0-nightly (6ef275e6c 2019-09-24)
binary: rustc
commit-hash: 6ef275e6c3cb1384ec78128eceeb4963ff788dca
commit-date: 2019-09-24
host: x86_64-unknown-linux-gnu
release: 1.39.0-nightly
LLVM version: 9.0
````

````rust
fn main() {
    const X: Vec<i32> = Vec::new();
    X.push(3);
    X.push(4);
    X.push(5);

    println!("{:?}", X);
}
````

This code looks like it should work, no compiler warning or error is generated.
The println will put out `[]` however.

If I replace `const` with `let mut`, I correctly get `[3,4,5]`.
Shouldn't `let mut` and `const` work the same way here?

**Comment by sfackler** [comment]
No - a constant is roughly like a `#define`. Your code is equivalent to

```rust
Vec::new().push(3);
Vec::new().push(4);
Vec::new().push(5);
```

Clippy has a lint for taking mutable references of constants for this kind of issue.

**Comment by Mark-Simulacrum** [comment]
Indeed; this is expected and not a bug -- I believe there are other issues that track the fact that we don't warn on 'useless code' so to speak here so I'm going to close (and it's not clear that we could -- technically, this code does quite a bit -- allocating and such).
